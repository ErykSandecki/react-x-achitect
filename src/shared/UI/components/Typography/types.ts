// others
import { COLORS } from 'constant/scss/variables/colors';

// types
import { Theme } from 'types/enums/theme';
import { TTypographyProps } from './Typography';
import { TValueObj } from 'types/generic';

export type TTypographyColorMode = {
  [Theme.dark]: TValueObj<
    Pick<
      typeof COLORS,
      | 'blue1Dark'
      | 'neutral1Dark'
      | 'neutral2Dark'
      | 'neutral3Dark'
      | 'neutral4Dark'
      | 'neutral5Dark'
    >
  >;
  [Theme.light]: TValueObj<
    Pick<
      typeof COLORS,
      | 'blue1Light'
      | 'neutral1Light'
      | 'neutral2Light'
      | 'neutral3Light'
      | 'neutral4Light'
      | 'neutral5Light'
    >
  >;
};

export type TTypographyColorsMode = {
  link: TTypographyColorMode;
  neutral1: TTypographyColorMode;
  neutral2: TTypographyColorMode;
  neutral3: TTypographyColorMode;
  neutral4: TTypographyColorMode;
  neutral5: TTypographyColorMode;
};

export type TTypographyColor = TTypographyColorMode | string;
export type TOmittedTypographyProps = Omit<TTypographyProps, 'fontType'>;
