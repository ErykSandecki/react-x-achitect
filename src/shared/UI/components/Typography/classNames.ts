import { mapValues } from 'lodash';

// types
import { ColorsTheme } from 'types/enums/scss/colorsTheme';
import {
  TypographyFontStyle,
  TypographyFontWeight,
  TypographyVariant,
} from './enums';

export const className = 'Typography';

export const classes = {
  className,
};

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      gutterBottom: `${className}--gutter-bottom`,
      noWrap: `${className}--no-wrap`,
      ...mapValues(ColorsTheme, (color) => `${className}--${color}`),
      ...mapValues(
        TypographyVariant,
        (fontType) => `${className}--${fontType}`,
      ),
      ...mapValues(
        TypographyFontStyle,
        (fontStyle) => `${className}--${fontStyle}`,
      ),
      ...mapValues(
        TypographyFontWeight,
        (fontWeight) => `${className}--${fontWeight}`,
      ),
    },
  },
};
