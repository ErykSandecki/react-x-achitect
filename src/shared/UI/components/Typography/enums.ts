export enum TypographyColor {
  blue1 = 'blue1',
  green1 = '28a745',
  green2 = 'a7b45d',
  neutral1 = 'neutral1',
  neutral2 = 'neutral2',
  neutral3 = 'neutral3',
  neutral4 = 'neutral4',
  neutral5 = 'neutral5',
  orange1 = 'orange1',
  orange2 = 'orange2',
  pink1 = 'pink1',
  red1 = 'red1',
  skyBlue1 = 'skyBlue1',
  violet1 = 'violet1',
}

export enum TypographyFontStyle {
  italic = 'italic',
  normal = 'normal',
}

export enum TypographyFontType {
  h1 = 'h1',
  h2 = 'h2',
  h3 = 'h3',
  h4 = 'h4',
  h5 = 'h5',
  h6 = 'h6',
  p = 'p',
  small = 'small',
}

export enum TypographyFontWeight {
  blacks = 'blacks',
  bold = 'bold',
  extraBold = 'extra-bold',
  light = 'light',
  medium = 'medium',
  regular = 'regular',
  semiBold = 'semi-bold',
}
