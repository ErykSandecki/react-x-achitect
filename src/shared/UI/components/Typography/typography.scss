@import '../../../../styles/mixins/text-ellipsis';
@import '../../../../styles/mixins/typography/typography';
@import '../../../../styles/mixins/typography/typography-list';
@import '../../../../styles/mixins/typography/typography-map';

.Typography {
  @each $color in $--rx-font-colors {
    &--#{$color} {
      @include theme($color, color);
    }
  }

  @each $font-type in $--rx-font-types {
    &--#{$font-type} {
      $--rx-font-size: map.get($--rx-font-size-map, $font-type);
      $--rx-line-height: map.get($--rx-line-height-map, $font-type);
      $--rx-margin-bottom: map.get($--rx-margin-bottom-map, $font-type);

      @include typography-styles(
        $--rx-font-size,
        $--rx-line-height,
        $--rx-margin-bottom
      );
    }
  }

  @each $font-style in $--rx-font-styles {
    &--#{$font-style} {
      font-style: map.get($--rx-font-styles-map, $font-style);
    }
  }

  @each $font-weight in $--rx-font-weights {
    &--#{$font-weight} {
      font-weight: map.get($--rx-font-weights-map, $font-weight);
    }
  }

  font-family: Rubik, sans-serif;

  &--no-wrap {
    @include text-ellipsis;
  }

  &--without-margin {
    margin: 0;
  }
}
