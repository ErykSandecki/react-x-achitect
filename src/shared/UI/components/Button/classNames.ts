import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonVariant } from './enums';
import { InputSize } from '../../enums';

export const className = 'Button';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      forcedHover: `${className}--forced-hover`,
      fullwidth: `${className}--full-width`,
      ...mapValues(ButtonColor, (color) => `${className}--${color}`),
      ...mapValues(InputSize, (size) => `${className}--${size}`),
      ...mapValues(ButtonVariant, (variant) => `${className}--${variant}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      applyFill: `${className}__icon--apply-fill`,
      applyStroke: `${className}__icon--apply-stroke`,
      end: `${className}__icon--end`,
      start: `${className}__icon--start`,
      ...mapValues(InputSize, (size) => `${className}__icon--${size}`),
    },
  },
};
