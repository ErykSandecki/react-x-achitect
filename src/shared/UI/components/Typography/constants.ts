// others
import { COLORS } from 'constant/scss/variables/colors';

// types
import { Theme } from 'types/enums/theme';
import { TTypographyColorsMode } from './types';

export const TYPOGRAPHY_COLORS_MODE: TTypographyColorsMode = {
  link: {
    [Theme.dark]: COLORS.blue1Dark,
    [Theme.light]: COLORS.blue1Light,
  },
  neutral1: {
    [Theme.dark]: COLORS.neutral1Dark,
    [Theme.light]: COLORS.neutral1Light,
  },
  neutral2: {
    [Theme.dark]: COLORS.neutral2Dark,
    [Theme.light]: COLORS.neutral2Light,
  },
  neutral3: {
    [Theme.dark]: COLORS.neutral3Dark,
    [Theme.light]: COLORS.neutral3Light,
  },
  neutral4: {
    [Theme.dark]: COLORS.neutral4Dark,
    [Theme.light]: COLORS.neutral4Light,
  },
  neutral5: {
    [Theme.dark]: COLORS.neutral5Dark,
    [Theme.light]: COLORS.neutral5Light,
  },
} as const;
