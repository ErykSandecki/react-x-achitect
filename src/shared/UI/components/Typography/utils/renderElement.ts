import { createElement, DetailedReactHTMLElement, JSX, ReactNode } from 'react';
import { isString } from 'lodash';

// types
import { TObject } from 'types';
import { TypographyVariant } from '../enums';
import { TTypograpghyComponent, TTypograpghyVariantMapping } from '../types';

export const renderElement = (
  children: ReactNode,
  component: TTypograpghyComponent | null,
  props: TObject<any>,
  variant: TypographyVariant,
  variantMapping: TTypograpghyVariantMapping,
): DetailedReactHTMLElement<TObject<any>, HTMLElement> | JSX.Element => {
  if (component) {
    return isString(component)
      ? createElement(component, props, children)
      : component({ children });
  }

  return createElement(variantMapping[variant] || variant, props, children);
};
