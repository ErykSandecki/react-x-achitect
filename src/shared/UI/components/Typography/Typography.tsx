import { camelCase, isObject } from 'lodash';
import {
  createElement,
  CSSProperties,
  forwardRef,
  HTMLAttributes,
  ReactNode,
  Ref,
} from 'react';

// hooks
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameTypography, classNames } from './classNames';
import { TYPOGRAPHY_COLORS_MODE } from './constants';

// styles
import styles from './typography.scss';

// types
import { E2EAttribute } from 'types/e2e';
import { TE2EDataAttributeProps } from '../../../E2EDataAttributes/E2EDataAttribute';
import { TTypographyColor } from './types';
import {
  TypographyFontStyle,
  TypographyFontWeight,
  TypographyFontType,
} from './enums';

// utils
import { getAttributes } from '../../../E2EDataAttributes/utils';

export type TTypographyProps = Omit<HTMLAttributes<HTMLElement>, 'color'> & {
  align?: CSSProperties['textAlign'];
  applyInlineColors?: boolean;
  children?: ReactNode;
  color?: TTypographyColor;
  e2eAttribute?: TE2EDataAttributeProps['type'];
  e2eValue?: TE2EDataAttributeProps['value'];
  fontStyle?: TypographyFontStyle;
  fontType?: TypographyFontType;
  fontWeight?: TypographyFontWeight;
  innerHtml?: string;
  noWrap?: boolean;
  ref?: Ref<HTMLElement>;
  withoutMargin?: boolean;
};

export const Typography = forwardRef<HTMLElement, TTypographyProps>(
  (
    {
      applyInlineColors = false,
      align: textAlign = 'inherit',
      children,
      className = '',
      color = TYPOGRAPHY_COLORS_MODE.neutral1,
      e2eAttribute = E2EAttribute.text,
      e2eValue = '',
      fontStyle = TypographyFontStyle.normal,
      fontType = TypographyFontType.p,
      fontWeight = TypographyFontWeight.regular,
      innerHtml = '',
      noWrap = false,
      style = {},
      withoutMargin = true,
      ...restProps
    },
    ref,
  ) => {
    const { classNamesWithTheme, cx, theme } = useTheme(classNames, styles);

    if (!children && !innerHtml) {
      return null;
    }

    return createElement(
      fontType,
      {
        ...restProps,
        className: cx(
          className,
          classNamesWithTheme[classNameTypography].name,
          classNamesWithTheme[classNameTypography].modificators[
            camelCase(fontWeight)
          ],
          classNamesWithTheme[classNameTypography].modificators[fontStyle],
          classNamesWithTheme[classNameTypography].modificators[fontType],
          [
            classNamesWithTheme[classNameTypography].modificators.noWrap,
            noWrap,
          ],
          [
            classNamesWithTheme[classNameTypography].modificators.withoutMargin,
            withoutMargin,
          ],
        ),
        ...getAttributes(e2eAttribute, e2eValue),
        ...(innerHtml
          ? { dangerouslySetInnerHTML: { __html: innerHtml } }
          : {}),
        ref,
        style: {
          ...style,
          ...(applyInlineColors
            ? { color: isObject(color) ? color[theme] : color }
            : {}),
          textAlign,
        },
      },
      children,
    );
  },
);

export default Typography;
