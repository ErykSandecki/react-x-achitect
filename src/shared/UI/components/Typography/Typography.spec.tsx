import { render } from '@testing-library/react';

// components
import Typography from './Typography';

// others
import { COLORS } from 'constant/scss/variables/colors';
import { className as classNameTypography, classNames } from './classNames';
import { TYPOGRAPHY_COLORS_MODE } from './constants';

// types
import { E2EAttribute } from 'types/e2e';
import {
  TypographyFontStyle,
  TypographyFontType,
  TypographyFontWeight,
} from './enums';

// utils
import { getByE2EAttribute } from 'test/testHelpers';
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';
import { enumToArray } from 'utils/transform/enumToArray';
import { hexToRgb } from '../../../../utils/transform/hexToRgb/hexToRgb';

const className = 'className';

describe('Typography props', () => {
  it('should pass align', () => {
    // mock
    const align = 'center';

    // before
    const { container } = render(
      <Typography align={align}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      'style',
      `text-align: ${align};`,
    );
  });

  it('should pass applyInlineColors', () => {
    // mock
    const rgb = hexToRgb(COLORS.neutral1Light);

    // before
    const { container } = render(
      <Typography applyInlineColors>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      'style',
      `color: rgb(${rgb.r}, ${rgb.g}, ${rgb.b}); text-align: inherit;`,
    );
  });

  it('should pass children', () => {
    // mock
    const children = 'children';

    //before
    const { container } = render(<Typography>{children}</Typography>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveTextContent(
      children,
    );
  });

  it('should pass className', () => {
    //before
    const { container } = render(
      <Typography className={className}>children</Typography>,
    );

    // result
    expect(container.querySelector(`.${className}`)).not.toBeNull();
  });

  it('should pass color', () => {
    //mock
    const rgb1 = hexToRgb(COLORS.blue1Light);
    const rgb2 = hexToRgb(COLORS.neutral5Light);

    // before
    const { container: container1 } = render(
      <Typography applyInlineColors color={COLORS.blue1Light}>
        children
      </Typography>,
    );

    const { container: container2 } = render(
      <Typography applyInlineColors color={TYPOGRAPHY_COLORS_MODE.neutral5}>
        children
      </Typography>,
    );

    // result
    expect(getByE2EAttribute(container1, E2EAttribute.text)).toHaveAttribute(
      'style',
      `color: rgb(${rgb1.r}, ${rgb1.g}, ${rgb1.b}); text-align: inherit;`,
    );

    expect(getByE2EAttribute(container2, E2EAttribute.text)).toHaveAttribute(
      'style',
      `color: rgb(${rgb2.r}, ${rgb2.g}, ${rgb2.b}); text-align: inherit;`,
    );
  });

  it('should pass e2eAttribute', () => {
    // before
    const { container } = render(
      <Typography e2eAttribute={E2EAttribute.text}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.text),
    );
  });

  it('should pass e2eValue', () => {
    // mock
    const e2eValue = 'e2eValue';

    // before
    const { container } = render(
      <Typography e2eValue={e2eValue}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.text),
      e2eValue,
    );
  });

  it('should pass fontStyle', () => {
    // mock
    const fontStyles = enumToArray<TypographyFontStyle>(TypographyFontStyle);

    //before
    const { container } = render(
      <>
        {fontStyles.map((fontStyle) => (
          <Typography
            e2eValue={fontStyle}
            fontStyle={fontStyle}
            key={fontStyle}
          >
            children
          </Typography>
        ))}
      </>,
    );

    // result
    fontStyles.forEach((fontStyle) => {
      expect(
        getByE2EAttribute(container, E2EAttribute.text, fontStyle),
      ).toHaveClass(classNames[classNameTypography].modificators[fontStyle]);
    });
  });

  it('should pass fontType', () => {
    // mock
    const fontTypes = enumToArray<TypographyFontType>(TypographyFontType);

    //before
    const { container } = render(
      <>
        {fontTypes.map((fontType) => (
          <Typography e2eValue={fontType} fontType={fontType} key={fontType}>
            children
          </Typography>
        ))}
      </>,
    );

    // result
    fontTypes.forEach((fontType) => {
      expect(
        getByE2EAttribute(container, E2EAttribute.text, fontType),
      ).toHaveClass(classNames[classNameTypography].modificators[fontType]);

      expect(container.getElementsByTagName(fontType)[0]).not.toBeNull();
    });
  });

  it('should pass fontWeight', () => {
    // mock
    const fontWeights = enumToArray<TypographyFontWeight>(TypographyFontWeight);

    //before
    const { container } = render(
      <>
        {fontWeights.map((fontWeight) => (
          <Typography
            e2eValue={fontWeight}
            fontWeight={fontWeight}
            key={fontWeight}
          >
            children
          </Typography>
        ))}
      </>,
    );

    // result
    fontWeights.forEach((fontWeight) => {
      expect(
        getByE2EAttribute(container, E2EAttribute.text, fontWeight),
      ).toHaveClass(classNames[classNameTypography].modificators[fontWeight]);
    });
  });

  it('should pass innerHtml', () => {
    // mock
    const children = 'children';
    const htmlTag = 'span';

    //before
    const { container } = render(
      <Typography innerHtml={`<${htmlTag}>${children}</${htmlTag}>`} />,
    );

    // result
    expect(container.getElementsByTagName(htmlTag)[0]).not.toBeNull();

    expect(container.getElementsByTagName(htmlTag)[0]).toHaveTextContent(
      children,
    );
  });

  it('should pass noWrap', () => {
    //before
    const { container } = render(<Typography noWrap>children</Typography>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveClass(
      classNames[classNameTypography].modificators.noWrap,
    );
  });

  it('should pass style', () => {
    // before
    const { container } = render(
      <Typography style={{ width: '100%' }}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).toHaveAttribute(
      'style',
      `width: 100%; text-align: inherit;`,
    );
  });

  it('should pass withoutMargin', () => {
    //before
    const { container } = render(
      <Typography withoutMargin={false}>children</Typography>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.text)).not.toHaveClass(
      classNames[classNameTypography].modificators.withoutMargin,
    );
  });
});

describe('Typography snapshots', () => {
  it('should render empty', () => {
    //before
    const { asFragment } = render(<Typography />);

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with some text', () => {
    // before
    const { asFragment } = render(<Typography>Text</Typography>);

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with passed inner html', () => {
    // before
    const { asFragment } = render(
      <Typography innerHtml={'<div>innerHtml</div>'} />,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
