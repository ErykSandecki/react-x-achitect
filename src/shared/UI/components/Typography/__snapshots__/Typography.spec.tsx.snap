// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography snapshots should render empty 1`] = `<DocumentFragment />`;

exports[`Typography snapshots should render with passed inner html 1`] = `
<DocumentFragment>
  <p
    class="Typography Typography--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--p Typography--p--light Typography--without-margin Typography--without-margin--light"
    data-test-text=""
    style="text-align: inherit;"
  />
  <div>
    innerHtml
  </div>
  <p />
</DocumentFragment>
`;

exports[`Typography snapshots should render with some text 1`] = `
<DocumentFragment>
  <p
    class="Typography Typography--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--p Typography--p--light Typography--without-margin Typography--without-margin--light"
    data-test-text=""
    style="text-align: inherit;"
  >
    Text
  </p>
</DocumentFragment>
`;
